tests/components/calendar/test_trigger.py

```diff
@@ -0,0 +1,375 @@
+"""Tests for the calendar automation.
+
+The tests create calendar based automations, set up a fake set of calendar
+events, then advance time to exercise that the automation is called. The
+tests use a fixture that mocks out events returned by the calendar entity,
+and create events using a relative time offset and then advance the clock
+forward exercising the triggers.
+"""
+from __future__ import annotations
+
+from collections.abc import Callable
+import datetime
+import logging
+import secrets
+from typing import Any, Generator
+from unittest.mock import patch
+
+import pytest
+
+from homeassistant.components import calendar
+import homeassistant.components.automation as automation
+from homeassistant.components.calendar.trigger import EVENT_START
+from homeassistant.core import HomeAssistant
+from homeassistant.setup import async_setup_component
+import homeassistant.util.dt as dt_util
+
+from tests.common import async_fire_time_changed, async_mock_service
+
+_LOGGER = logging.getLogger(__name__)
+
+
+CALENDAR_ENTITY_ID = "calendar.calendar_2"
+CONFIG = {calendar.DOMAIN: {"platform": "demo"}}
+
+TEST_AUTOMATION_ACTION = {
+    "service": "test.automation",
+    "data_template": {
+        "platform": "{{ trigger.platform }}",
+        "event": "{{ trigger.event }}",
+        "calendar_event": "{{ trigger.calendar_event }}",
+    },
+}
+
+# The trigger sets two alarms: One based on the next event and one
+# to refresh the schedule. The test advances the time an arbitrary
+# amount to trigger either type of event with a small jitter.
+TEST_TIME_ADVANCE_INTERVAL = datetime.timedelta(minutes=1)
+TEST_UPDATE_INTERVAL = datetime.timedelta(minutes=15)
+
+
+@pytest.fixture
+async def now() -> datetime.datetime:
+    """Fixture to provide a base time for tests."""
+    return dt_util.utcnow()
```

Comment:
Consider using freezegun instead

Reply:
I am familiar with freezegun for the basic use in a basic test, though having a little trouble knowing the best way to use it for these fixtures.

Right now this test is being a little tricky in that We set a fixed time, then fire fake alarms in the future with wide enough windows that we assume we won't ever hit flakiness while the clock continue to advance. I like the idea of advancing explicitly with freezegun instead, so i agree there is potential here.

For a fixture like `fake_schedule`: should each test freeze its own time and then the fixture calls "now" knowing its getting a frozen time? or just throw a `@pytest.mark.freeze_time` on the `now` fixture, but still advance explicitly during tests with tick? A little more detail, or pointer to an example might help get me moving on it.

Reply:
Figured out something here, done.